platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/isolated/IsolatedProjectsToolingApiBuildActionIntegrationTest.groovy

```diff
@@ -487,4 +489,136 @@ class IsolatedProjectsToolingApiBuildActionIntegrationTest extends AbstractIsola
         fixture.assertStateLoaded()
     }
 
+    def "caches execution of different BuildAction types"() {
+        given:
+        withSomeToolingModelBuilderPluginInBuildSrc()
+        buildFile << """
+            plugins.apply(my.MyPlugin)
+        """
+
+        when:
+        executer.withArguments(ENABLE_CLI)
+        def model = runBuildAction(new FetchCustomModelForEachProject())
+
+        then:
+        fixture.assertStateStored {
+            projectConfigured(":buildSrc")
+            buildModelCreated()
+            modelsCreated(":")
+        }
+        outputContains("creating model for root project 'root'")
+
+        and:
+        model.size() == 1
+        model[0].message == "It works from project :"
+
+
+        when:
+        executer.withArguments(ENABLE_CLI)
+        def model2 = runBuildAction(new FetchModelsMultipleTimesForEachProject())
+
+        then:
+        fixture.assertStateStored {
+            projectConfigured(":buildSrc")
+            buildModelCreated()
+            modelsCreated(":")
+        }
+        outputContains("creating model for root project 'root'")
+
+        and:
+        model2.size() == 2
+        model2[0].message == "It works from project :"
+
+
+        when:
+        executer.withArguments(ENABLE_CLI)
+        def model3 = runBuildAction(new FetchCustomModelForEachProject())
+
+        then:
+        fixture.assertStateLoaded()
```

Comment:
❓ Is this enough to assert that no models had to be recreated?

Reply:
The result of executing a build action is stored in the configuration cache. If it was loaded, then the provided build action has not been executed. This implies that no models were created to provide the result.

Reply:
Oh, and it also explicitly asserts that there were no build operations for model queries 🙃 

https://github.com/gradle/gradle/blob/master/platforms/core-configuration/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/isolated/IsolatedProjectsFixture.groovy#L166